common/sensor/dev/pex89000.c

```diff
@@ -0,0 +1,353 @@
+	memcpy(&msg.data[0], &cmd, sizeof(cmd));
+
+	if (i2c_master_read(&msg, retry)) {
+		/* read fail */
+		printf("pex89000 read failed!!\n");
+		return 0;
+	}
+
+	memcpy(resp, &msg.data[0], resp_len);
+
+	return 1;
+}
+
+static uint8_t pex89000_chime_write(uint8_t bus, uint8_t addr, uint32_t oft, uint8_t *data,
+				    uint8_t data_len)
+{
+	if (!set_full_addr(bus, addr, oft))
+		return 0;
+
+	HW_I2C_Cmd cmd;
+	pex89000_i2c_encode(oft, 0xF, BRCM_I2C5_CMD_WRITE, BRCM_I2C5_ACCESS_MODE_FULL, &cmd);
+
+	/* This buffer is for PEC calculated and write to switch by I2C */
+	uint8_t *data_buf = (uint8_t *)malloc(sizeof(cmd) + data_len);
+	if (!data_buf)
+		return 0;
+	memcpy(data_buf, &cmd, sizeof(cmd));
+	memcpy(data_buf + sizeof(cmd), data, data_len);
+
+	uint8_t retry = 5;
+	I2C_MSG msg;
+
+	msg.bus = bus;
+	msg.slave_addr = addr;
+	msg.tx_len = sizeof(cmd) + data_len;
+	memcpy(&msg.data[0], data_buf, sizeof(cmd) + data_len);
+
+	if (i2c_master_write(&msg, retry)) {
+		/* write fail */
+		printf("pex89000 write failed!!\n");
+		free(data_buf);
+		return 0;
+	}
+
+	free(data_buf);
+	return 1;
+}
+
+static uint8_t pend_for_read_valid(uint8_t bus, uint8_t addr)
+{
+	uint8_t rty = 50;
+	uint32_t resp = 0;
+
+	while (rty--) {
+		if (!pex89000_chime_read(bus, addr, BRCM_CHIME_AXI_CSR_CTL, (uint8_t *)&resp,
+					 sizeof(resp))) {
+			k_msleep(10);
+			continue;
+		}
+
+		if ((resp >> 24) & 0x8) { // CHIME_to_AXI_CSR Control Status -> Read_data_vaild
+			return 1; //  success
+		}
+
+		k_msleep(10);
+	}
+
+	return 0;
+}
+
+static uint8_t pex89000_chime_to_axi_write(uint8_t bus, uint8_t addr, uint32_t oft, uint32_t data)
+{
+	uint8_t rc = 0;
+
+	uint32_t wbuf = oft;
+
+	swap32(&wbuf);
+	if (!pex89000_chime_write(bus, addr, BRCM_CHIME_AXI_CSR_ADDR, (uint8_t *)&wbuf,
+				  sizeof(wbuf))) {
+		goto exit;
+	}
+	wbuf = data;
+	swap32(&wbuf);
+	if (!pex89000_chime_write(bus, addr, BRCM_CHIME_AXI_CSR_DATA, (uint8_t *)&wbuf,
+				  sizeof(wbuf))) {
+		goto exit;
+	}
+	wbuf = 0x1; // CHIME_to_AXI_CSR Control Status: write command
+	swap32(&wbuf);
+	if (!pex89000_chime_write(bus, addr, BRCM_CHIME_AXI_CSR_CTL, (uint8_t *)&wbuf,
+				  sizeof(wbuf))) {
+		goto exit;
+	}
+
+	rc = 1;
+
+exit:
+	return rc;
+}
+
+static uint8_t pex89000_chime_to_axi_read(uint8_t bus, uint8_t addr, uint32_t oft, uint32_t *resp)
+{
+	uint8_t rc = 0;
+
+	uint32_t data = oft;
+	swap32(&data);
+	if (!pex89000_chime_write(bus, addr, BRCM_CHIME_AXI_CSR_ADDR, (uint8_t *)&data,
+				  sizeof(data))) {
+		goto exit;
+	}
+	data = 0x2; // CHIME_to_AXI_CSR Control Status: read command
+	swap32(&data);
+	if (!pex89000_chime_write(bus, addr, BRCM_CHIME_AXI_CSR_CTL, (uint8_t *)&data,
+				  sizeof(data))) {
+		goto exit;
+	}
+
+	k_msleep(10);
+	if (!pend_for_read_valid(bus, addr)) {
+		printf("read data invaild\n");
+		goto exit;
+	}
+
+	if (!pex89000_chime_read(bus, addr, BRCM_CHIME_AXI_CSR_DATA, (uint8_t *)resp,
+				 sizeof(resp))) {
+		goto exit;
+	}
+
+	swap32(resp);
+	rc = 1;
+
+exit:
+	return rc;
```

Comment:
Same here, goto seems unnecessary because there isn't any cleanup happening as far as I can tell.
Just return error when it happens.